### PR TITLE
Return devices

### PR DESCRIPTION
**Pull request by johns10**
I want to ship this in a desktop app. So, I'd like to call `Membrane.PortAudio.Devices.list()` and get back a list of devices so the user can select one.

I implemented it by adding a `device` spec and iterating over the DeviceInfo structs to create `devices`. I then return those in a list. 

This is my first crack at NIF's so be gentle.

### CODE REVIEW

**Comment by thbar** [comment]
I was about to start a PR on exactly this, since I have the exact same need (I'm capturing sound cards audio, which can get plugged in / out, not on the same port etc).

Thanks for your work!

For history context:
- https://github.com/membraneframework/membrane_portaudio_plugin/issues/9
- https://github.com/membraneframework/membrane_portaudio_plugin/pull/43

**Review by mat-hek — CHANGES_REQUESTED** [review]

**Comment by johns10** [comment]
I'll hit these changes shortly.

File: c_src/membrane_portaudio_plugin/pa_devices.c [hunk 1]
```diff
@@ -1,31 +1,48 @@
...
           i == default_input_id
               ? " (default input)"
-              : i == default_output_id ? " (default output)" : "";
+          : i == default_output_id ? " (default output)"
+                                   : "";
 
       printf("%s%s\r\n\tid: %d\r\n\tmax_input_channels: "
              "%d\r\n\tmax_output_channels: %d\r\n\n",
              device_info->name, default_str, i, device_info->maxInputChannels,
              device_info->maxOutputChannels);
```
[the remark this answers is not in the document]
Comment (johns10) [reply c_src/membrane_portaudio_plugin/pa_devices.c:35]: Agreed, and done.

File: c_src/membrane_portaudio_plugin/pa_devices.spec.exs [hunk 2]
```diff
@@ -1,3 +1,12 @@
...
-spec list() :: :ok
+type(
+  device :: %Device{
+    id: int,
+    name: string,
+    max_input_channels: int,
+    max_output_channels: int
+  }
+)
+
```
[the remark this answers is not in the document]
Comment (johns10) [reply c_src/membrane_portaudio_plugin/pa_devices.spec.exs:11]: I had added is_default, will move to this instead.

**Comment by johns10** [comment]
Thanks for considering my changes, and your valuable feedback. 

I implemented everything you've recommended.

I do question the way we are handling returns.
There are some error cases here.
I don't know Unifex enough to understand how and when to return the ok and error tuples.

Since I removed the :ok from the spec, I can no longer use return_list_ok. 
Of course, we can just return empty lists in our error case.
Would it be better for us to return `{:ok, devices}` or `{:error, "Pa_CountDevices returned #{numDevices"}` to retain the semantics of the model?

Given my limited experience with Unifex, it's not obvious how to do this, so I just accepted your proposal.
If you think this would be beneficial, I can attempt a refactor. 
Your input on syntax would be very helpful.

Additionally, I was seeing intermittent segmentation faults in dev, so I changed the way I was assigning names to dynamically allocate memory on the device, and then copy the value from deviceInfo. I hope that resolves the segmentation fault, but for the safety of your users I should probably kick this around on my local a bit before you merge this PR.

Maybe give it a closer eye and make sure I haven't done anything grotesque with memory management? I'm a bit green with C.

**Comment by mat-hek** [comment]
Hi @johns10, basically, the `label` in `.spec.exs` maps to the result function name. Have you seen https://hexdocs.pm/unifex/Unifex.Specs.DSL.html? There's also https://hexdocs.pm/unifex/creating_unifex_natives.html. Let me know if we can improve the docs somehow to make it clearer.

Regarding this particular error, you're right, we need to do something about it. If we're about to return an error, it should be somehow structured, for example `{:error, {:init, error_no}}` and then we should leave the `:ok` as well. However, I'd rather raise with `unifex_raise(env, description)`.

File: c_src/membrane_portaudio_plugin/pa_devices.c [hunk 3]
```diff
@@ -1,31 +1,52 @@
...
+      else
+      {
+        devices[i].default_device = DEFAULT_DEVICE_FALSE;
+      }
+
+      devices[i].name = malloc(strlen(device_info->name) + 1);
+      if (device_info->name != NULL)
+      {
+        strcpy(devices[i].name, device_info->name);
+      }
```
Comment (mat-hek) [inline c_src/membrane_portaudio_plugin/pa_devices.c:42]: If you allocate anything on the heap, you should free it. In this case, after the call to `list_result`. However, unifex does malloc and memcpy when converting to an erlang term, so it seems redundant.

File: lib/membrane_portaudio_plugin/device.ex [hunk 4]
```diff
@@ -0,0 +1,10 @@
+defmodule Membrane.PortAudio.Device do
```
Comment (mat-hek) [inline lib/membrane_portaudio_plugin/device.ex:1]: a piece of doc would be great

File: lib/membrane_portaudio.ex [hunk 5]
```diff
@@ -20,6 +20,8 @@ defmodule Membrane.PortAudio do
   @spec print_devices() :: :ok
   def print_devices() do
     Application.ensure_all_started(:membrane_portaudio_plugin)
+
     __MODULE__.SyncExecutor.apply(__MODULE__.Devices, :list, [])
+    |> IO.inspect()
```
Comment (mat-hek) [inline lib/membrane_portaudio.ex:25]: Let's add `list_devices/0` here too, the existing one is in a private module and not guarded with the sync executor

**Review by mat-hek — CHANGES_REQUESTED** [review]

**Pull request merged**